sdk/storage/inc/datalake/protocol/datalake_rest_client.hpp

```diff
@@ -0,0 +1,2244 @@
+  constexpr static const char* k_HeaderXMsLeaseStatus = "x-ms-lease-status";
+  constexpr static const char* k_HeaderXMsContentMd5 = "x-ms-content-md5";
+  constexpr static const char* k_HeaderXMsOwner = "x-ms-owner";
+  constexpr static const char* k_HeaderXMsGroup = "x-ms-group";
+  constexpr static const char* k_HeaderXMsPermissions = "x-ms-permissions";
+  constexpr static const char* k_HeaderXMsAcl = "x-ms-acl";
+  constexpr static const char* k_HeaderXMsClientRequestId = "x-ms-client-request-id";
+
+  // Mode "set" sets POSIX access control rights on files and directories, "modify" modifies one or more POSIX access control rights  that pre-exist on files and directories, "remove" removes one or more POSIX access control rights  that were present earlier on files and directories
+  enum class PathSetAccessControlRecursiveMode
+  {
+    Set,
+    Modify,
+    Remove,
+    Unknown
+  };
+
+  std::string PathSetAccessControlRecursiveModeToString(const PathSetAccessControlRecursiveMode& pathSetAccessControlRecursiveMode)
+  {
+    switch (pathSetAccessControlRecursiveMode)
+    {
+      case PathSetAccessControlRecursiveMode::Set:
+        return "set";
+      case PathSetAccessControlRecursiveMode::Modify:
+        return "modify";
+      case PathSetAccessControlRecursiveMode::Remove:
+        return "remove";
+      default:
+        return std::string();
+    }
+  }
+
+  PathSetAccessControlRecursiveMode PathSetAccessControlRecursiveModeFromString(const std::string& pathSetAccessControlRecursiveMode)
+  {
+    if (pathSetAccessControlRecursiveMode == "set")
+    {
+      return PathSetAccessControlRecursiveMode::Set;
+    }
+    if (pathSetAccessControlRecursiveMode == "modify")
+    {
+      return PathSetAccessControlRecursiveMode::Modify;
+    }
+    if (pathSetAccessControlRecursiveMode == "remove")
+    {
+      return PathSetAccessControlRecursiveMode::Remove;
+    }
+    throw std::runtime_error("Cannot convert " + pathSetAccessControlRecursiveMode + " to PathSetAccessControlRecursiveMode");
+  };
+
+  struct AclFailedEntry
+  {
+    std::string Name;
+    std::string Type;
+    std::string ErrorMessage;
+
+    static AclFailedEntry CreateFromJson(const nlohmann::json& node)
+    {
+      AclFailedEntry result;
+      result.Name = node["name"].get<std::string>();
+      result.Type = node["type"].get<std::string>();
+      result.ErrorMessage = node["errorMessage"].get<std::string>();
+      return result;
+    }
+  };
+
+  struct SetAccessControlRecursiveResponse
+  {
+    uint32_t DirectoriesSuccessful;
+    uint32_t FilesSuccessful;
+    uint32_t FailureCount;
+    std::vector<AclFailedEntry> FailedEntries;
+
+    static SetAccessControlRecursiveResponse CreateFromJson(const nlohmann::json& node)
+    {
+      SetAccessControlRecursiveResponse result;
+      result.DirectoriesSuccessful = static_cast<uint32_t>(std::stoul(node["directoriesSuccessful"].get<std::string>()));
```

Comment:
`std::stoul` returns `unsigned long`, which is 4 bytes on 32 bit platforms and 8 bytes on 64 bit platforms. I think this is something we want to avoid as a portable library.

Reply:
I also thought about this. The value itself is always at most 4 bytes, so it is safe to cast it down.
Unfortunately, there isn't an std function to convert the string directly to `uint32_t` so we have to use another function here.
An alternative is to use `std::stoull` on all platforms and cast it down to `uint32_t`, but I don't prefer this one.